File: frontend/src/ScoringWeights.test.ts
import { describe, it, expect } from 'vitest';
import { ScoringWeights } from './types';

describe('ScoringWeights Validation Logic', () => {
  it('should validate standard default weights sum to exactly 100', () => {
    const defaultWeights: ScoringWeights = {
      skill: 30,
      experience: 15,
      visa: 20,
      location: 10,
      salary: 10,
      career: 10,
      difficulty: 5,
    };

    const sum = Object.values(defaultWeights).reduce((a, b) => a + b, 0);
    expect(sum).toBe(100);
  });

  it('should detect when weights do not sum to 100', () => {
    const invalidWeights: ScoringWeights = {
      skill: 30,
      experience: 15,
      visa: 20,
      location: 10,
      salary: 10,
      career: 10,
      difficulty: 4, // 99 total
    };

    const sum = Object.values(invalidWeights).reduce((a, b) => a + b, 0);
    expect(sum).not.toBe(100);
    expect(sum).toBe(99);
  });
});

File: backend/src/test/java/com/personal/jobagent/preferences/ScoringWeightsTest.java
package com.personal.jobagent.preferences;

import org.junit.jupiter.api.Test;

import java.util.HashMap;
import java.util.Map;

import static org.assertj.core.api.Assertions.assertThatCode;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

class ScoringWeightsTest {

    @Test
    void validWeightsSummingTo100Passes() {
        Map<String, Object> weights = Map.of(
                "skill", 30,
                "experience", 15,
                "visa", 20,
                "location", 10,
                "salary", 10,
                "career", 10,
                "difficulty", 5
        );

        assertThatCode(() -> ScoringWeights.validate(weights)).doesNotThrowAnyException();
    }

    @Test
    void missingCategoryThrows() {
        Map<String, Object> weights = Map.of(
                "skill", 35,
                "experience", 15,
                "visa", 20,
                "location", 10,
                "salary", 10,
                "career", 10
                // missing difficulty
        );

        assertThatThrownBy(() -> ScoringWeights.validate(weights))
                .isInstanceOf(ScoringWeights.ValidationException.class)
                .hasMessageContaining("must contain exactly these categories");
    }

    @Test
    void sumNot100Throws() {
        Map<String, Object> weights = Map.of(
                "skill", 30,
                "experience", 15,
                "visa", 20,
                "location", 10,
                "salary", 10,
                "career", 10,
                "difficulty", 4 // sum = 99
        );

        assertThatThrownBy(() -> ScoringWeights.validate(weights))
                .isInstanceOf(ScoringWeights.ValidationException.class)
                .hasMessageContaining("must sum to 100, got 99");
    }

    @Test
    void negativeWeightThrows() {
        Map<String, Object> weights = Map.of(
                "skill", 35,
                "experience", 15,
                "visa", 20,
                "location", 10,
                "salary", 15,
                "career", 10,
                "difficulty", -5 // sum = 100 but negative
        );

        assertThatThrownBy(() -> ScoringWeights.validate(weights))
                .isInstanceOf(ScoringWeights.ValidationException.class)
                .hasMessageContaining("must not be negative");
    }

    @Test
    void nullThrows() {
        assertThatThrownBy(() -> ScoringWeights.validate(null))
                .isInstanceOf(ScoringWeights.ValidationException.class)
                .hasMessageContaining("scoringWeights is required");
    }
}
